sdk/core/azure-core/src/http/curl/curl_connection_pool_private.hpp

```diff
@@ -90,51 +87,47 @@ namespace Azure { namespace Core { namespace Http { namespace _detail {
         std::unique_ptr<CurlNetworkConnection> connection,
         HttpStatusCode lastStatusCode);
 
-    // Class can't have instances.
-    CurlConnectionPool() = delete;
-
-    static void StopCleaner() { g_isCleanConnectionsRunning = false; }
+    /**
+     * @brief Remove all connections from the pool.
+     *
+     */
+    void ResetPool() { ConnectionPoolIndex.clear(); }
 
-  private:
     /**
-     * Review all connections in the pool and removes old connections that might be already
-     * expired and closed its connection on server side.
+     * @brief Keeps a unique key for each host and creates a connection pool for each key.
+     *
+     * @details This way getting a connection for a specific host can be done in O(1) instead of
+     * looping a single connection list to find the first connection for the required host.
+     *
+     * @remark There might be multiple connections for each host.
      */
-    static void CleanUp();
+    std::map<std::string, std::list<std::unique_ptr<CurlNetworkConnection>>> ConnectionPoolIndex;
 
-    AZ_CORE_DLLEXPORT static uint64_t g_connectionCounter;
-    AZ_CORE_DLLEXPORT static std::atomic<bool> g_isCleanConnectionsRunning;
-    // Removes all connections and indexes
-    static void ClearIndex() { CurlConnectionPool::ConnectionPoolIndex.clear(); }
+    std::mutex ConnectionPoolMutex;
 
-    // Makes possible to know the number of current connections in the connection pool for an
-    // index
-    static int64_t ConnectionsOnPool(std::string const& host)
-    {
-      auto& pool = CurlConnectionPool::ConnectionPoolIndex[host];
-      return pool.size();
-    };
+    // This is used to put the cleaning pool thread to sleep and yet to be able to wake it if the
+    // application finishes.
+    std::condition_variable ConditionalVariableForCleanThread;
 
-    // Makes possible to know the number indexes in the pool
-    static int64_t ConnectionsIndexOnPool()
-    {
-      return CurlConnectionPool::ConnectionPoolIndex.size();
-    };
-  };
+    AZ_CORE_DLLEXPORT static Azure::Core::Http::_detail::CurlConnectionPool g_curlConnectionPool;
 
-  struct CurlGlobalStateForAzureSdk
-  {
-    CurlGlobalStateForAzureSdk() { curl_global_init(CURL_GLOBAL_ALL); }
+  private:
+    // private constructor to keep this as singleton.
+    CurlConnectionPool() { curl_global_init(CURL_GLOBAL_ALL); }
 
-    ~CurlGlobalStateForAzureSdk()
-    {
-      CurlConnectionPool::StopCleaner();
-      curl_global_cleanup();
-    }
+    std::thread m_cleanThread;
+
+    // Makes possible to know the number of current connections in the connection pool for an
+    // index
+    int64_t ConnectionsOnPool(std::string const& host) { return ConnectionPoolIndex[host].size(); };
+
+    // Define the maximun allowed connections per host-index in the pool. If this number is reached
+    // for the host-index, next connections trying to be added to the pool will be ignored.
+    uint64_t m_maxConnectionsPerIndex = 1024;
```

Comment:
Make this const? Don't we typically use a PascalCase name for similar constants?

For example:
https://github.com/Azure/azure-sdk-for-cpp/blob/0b7201d6ccc10f369beccc15db183f51ac3d0eb2/sdk/core/azure-core/inc/azure/core/http/win_http_transport.hpp#L23

Reply:
In this case this is just a private member of the class.  It's OK IMO because the class is a singleton. The member becomes part of an static instance.

I think I can also let it out next to all other constants.  Since it is only used when a connection is moved back to the pool, I don't need to worry about this constant being destroyed on app exit before some thread tries to access it